src/bin/elfshaker/utils.rs

```diff
@@ -107,23 +109,126 @@ pub fn open_repo_from_cwd() -> Result<Repository, RepoError> {
+        is_terminal_escape_supported(std::io::stdout());
+}
+
+/// Clears the current line
+const ESC_CLEAR_LINE: &str = "\x1b[2K";
+const ESC_BOLD_ON: &str = "\x1b[1m";
+const ESC_GREEN: &str = "\x1b[32;1m";
+const ESC_CYAN: &str = "\x1b[36;1m";
+/// Resets all text formatting
+const ESC_RESET: &str = "\x1b[32;0m";
+
 pub fn create_percentage_print_reporter(message: &str, step: u32) -> ProgressReporter<'static> {
     assert!(step <= 100);
 
     let current = Arc::new(AtomicIsize::new(-100));
     let message = message.to_owned();
     ProgressReporter::new(move |checkpoint| {
         if let Some(remaining) = checkpoint.remaining {
-            let percentage =
-                (100 * checkpoint.done / std::cmp::max(1, checkpoint.done + remaining)) as isize;
-            if percentage - current.load(Ordering::Acquire) >= step as isize {
+            let is_terminal_escape_supported = *STDOUT_TERMINAL_ESCAPE_SUPPORTED;
+            // Helper which returns the input if terminal codes are supported
+            let if_escape = |s| {
+                if is_terminal_escape_supported {
+                    s
+                } else {
+                    ""
+                }
+            };
+
+            let total = checkpoint.done + remaining;
+            let percentage = (100 * checkpoint.done / std::cmp::max(1, total)) as isize;
+
+            // Stepping helps to avoid too many lines being printed,
+            // but when using terminal escape codes, we refresh the same
+            // line over an over, so it is not needed.
+            let needs_update = percentage - current.load(Ordering::Acquire) >= step as isize;
+            if needs_update || is_terminal_escape_supported {
                 current.store(percentage, Ordering::Release);
-                eprintln!("{}... {}%", message, percentage);
+                // Clear line and reset cursor
+                print!("{}\r", if_escape(ESC_CLEAR_LINE));
+
+                if percentage == 100 {
+                    // Bold green success text
+                    println!(
+                        "{}{}{}{} [{}/{}]",
+                        if_escape(ESC_BOLD_ON),
+                        if_escape(ESC_GREEN),
+                        message,
+                        if_escape(ESC_RESET),
+                        total,
+                        total
+                    );
+                } else {
+                    // Bold blue status text with percentage
+                    let status = format!(
+                        "{}{}{}{} {} {}% [{}/{}]",
+                        if_escape(ESC_BOLD_ON),
+                        if_escape(ESC_CYAN),
+                        message,
+                        if_escape(ESC_RESET),
+                        create_progress_bar(percentage as i32, DEFAULT_PROGRESS_BAR_WIDTH),
+                        percentage,
+                        checkpoint.done,
+                        total,
+                    );
```

Comment:
It's a small nit, but is format preferred over a straight print here, if it can be rearranged? I'm thinking that format may allocate, which isn't free, whereas print doesn't have to (though it may).

Reply:
Done to avoid code duplication, so not necessary. I think that I/O would trump allocating a short string, but I agree that there is no reason not to make this optimisation.